File: js/webgpu/internal/machine-prep.mjs
/**
 * Machine preparation: JSON machine + params -> dense log_trans tensor.
 *
 * Converts a Machine Boss JSON transducer into the dense representation
 * needed by DP algorithms: log_trans[in_tok][out_tok][src][dst] as a
 * flat Float32Array (for GPU) or Float64Array (for CPU).
 *
 * Token convention: token 0 is the "null" (epsilon/gap) token.
 * Real tokens are 1-based indices into the sorted alphabet.
 */

import { NEG_INF } from './logmath.mjs';

/**
 * Evaluate a weight expression to a float given parameter values.
 * Mirrors python/machineboss/weight.py:evaluate().
 *
 * @param {*} w - Weight expression (number, string, or object)
 * @param {Object<string,number>} params - Parameter name -> value
 * @param {Object<string,*>} [defs] - Function/parameter definitions from machine
 * @returns {number}
 */
export function evaluateWeight(w, params, defs) {
  if (w === null || w === undefined) return 0;
  if (typeof w === 'number') return w;
  if (typeof w === 'boolean') return w ? 1.0 : 0.0;
  if (typeof w === 'string') {
    if (params != null && w in params) return params[w];
    if (defs != null && w in defs) return evaluateWeight(defs[w], params, defs);
    throw new Error(`Unknown parameter "${w}". Pass it in the params object, e.g. { ${w}: 0.5 }`);
  }
  if (typeof w === 'object') {
    if ('*' in w) {
      const [a, b] = w['*'];
      return evaluateWeight(a, params, defs) * evaluateWeight(b, params, defs);
    }
    if ('+' in w) {
      const [a, b] = w['+'];
      return evaluateWeight(a, params, defs) + evaluateWeight(b, params, defs);
    }
    if ('-' in w) {
      const [a, b] = w['-'];
      return evaluateWeight(a, params, defs) - evaluateWeight(b, params, defs);
    }
    if ('/' in w) {
      const [a, b] = w['/'];
      return evaluateWeight(a, params, defs) / evaluateWeight(b, params, defs);
    }
    if ('pow' in w) {
      const [base, exp] = w['pow'];
      return Math.pow(evaluateWeight(base, params, defs), evaluateWeight(exp, params, defs));
    }
    if ('log' in w) {
      return Math.log(evaluateWeight(w['log'], params, defs));
    }
    if ('exp' in w) {
      return Math.exp(evaluateWeight(w['exp'], params, defs));
    }
    if ('not' in w) {
      return 1.0 - evaluateWeight(w['not'], params, defs);
    }
    throw new Error(`Unsupported weight operator "${Object.keys(w).join(', ')}". Supported: *, +, -, /, pow, log, exp, not`);
  }
  throw new TypeError(`Unsupported weight expression type: ${typeof w}`);
}

/**
 * Prepared machine: dense transition tensor + metadata.
 * @typedef {Object} PreparedMachine
 * @property {number} nStates - Number of states (S)
 * @property {number} nInputTokens - Number of input tokens including null (n_in)
 * @property {number} nOutputTokens - Number of output tokens including null (n_out)
 * @property {string[]} inputAlphabet - Input token symbols (index 0 = null)
 * @property {string[]} outputAlphabet - Output token symbols (index 0 = null)
 * @property {Float64Array} logTrans - Dense tensor (n_in * n_out * S * S), row-major
 * @property {Float32Array} logTransF32 - Same as logTrans but Float32 for GPU
 */

/**
 * Build sorted token alphabet from machine transitions.
 * Token 0 is reserved for null (epsilon). Real tokens are 1-based.
 *
 * @param {Array} states - Machine JSON states array
 * @param {'in'|'out'} direction
 * @returns {string[]} alphabet where index 0 is null string
 */
function buildAlphabet(states, direction) {
  const tokens = new Set();
  for (const state of states) {
    for (const trans of (state.trans || [])) {
      const tok = trans[direction];
      if (tok) tokens.add(tok);
    }
  }
  const sorted = Array.from(tokens).sort();
  return ['', ...sorted];  // index 0 = null/epsilon
}

/**
 * Prepare a machine JSON + params into a dense log transition tensor.
 *
 * @param {Object} machineJSON - Machine Boss JSON object
 * @param {Object<string,number>} [params={}] - Parameter values
 * @returns {PreparedMachine}
 */
export function prepareMachine(machineJSON, params = {}) {
  const states = machineJSON.state;
  const defs = machineJSON.defs || {};
  const S = states.length;

  const inputAlphabet = buildAlphabet(states, 'in');
  const outputAlphabet = buildAlphabet(states, 'out');
  const nIn = inputAlphabet.length;
  const nOut = outputAlphabet.length;

  // Build token -> index maps
  const inTokIdx = {};
  for (let i = 0; i < nIn; i++) inTokIdx[inputAlphabet[i]] = i;
  const outTokIdx = {};
  for (let i = 0; i < nOut; i++) outTokIdx[outputAlphabet[i]] = i;

  // Resolve state name references to indices
  const nameToIdx = {};
  for (let i = 0; i < S; i++) {
    const name = states[i].id;
    if (name !== undefined) {
      const key = Array.isArray(name) ? JSON.stringify(name) : name;
      nameToIdx[key] = i;
    }
    nameToIdx[i] = i;
  }

  function resolveDest(dest) {
    if (typeof dest === 'number') return dest;
    const key = Array.isArray(dest) ? JSON.stringify(dest) : dest;
    if (key in nameToIdx) return nameToIdx[key];
    throw new Error(`Transition references unknown state "${dest}". Check that the "to" field matches a state "id" in your machine JSON.`);
  }

  // Allocate dense tensor: logTrans[inTok * nOut * S * S + outTok * S * S + src * S + dst]
  const size = nIn * nOut * S * S;
  const logTrans = new Float64Array(size).fill(NEG_INF);

  // Fill from transitions
  for (let src = 0; src < S; src++) {
    const trans = states[src].trans || [];
    for (const t of trans) {
      const dst = resolveDest(t.to);
      const inIdx = t.in ? inTokIdx[t.in] : 0;
      const outIdx = t.out ? outTokIdx[t.out] : 0;
      const logWeight = Math.log(evaluateWeight(t.weight != null ? t.weight : 1, params, defs));
      const idx = ((inIdx * nOut + outIdx) * S + src) * S + dst;
      // logaddexp in case there are duplicate transitions
      if (logTrans[idx] === NEG_INF) {
        logTrans[idx] = logWeight;
      } else {
        const a = logTrans[idx], b = logWeight;
        const m = a > b ? a : b;
        logTrans[idx] = m + Math.log(Math.exp(a - m) + Math.exp(b - m));
      }
    }
  }

  // Float32 version for GPU
  const logTransF32 = new Float32Array(size);
  for (let i = 0; i < size; i++) {
    logTransF32[i] = logTrans[i] === NEG_INF ? -Infinity : logTrans[i];
  }

  return {
    nStates: S,
    nInputTokens: nIn,
    nOutputTokens: nOut,
    inputAlphabet,
    outputAlphabet,
    logTrans,
    logTransF32,
  };
}

/**
 * Look up token index for a symbol in an alphabet.
 * @param {string[]} alphabet
 * @param {string} symbol
 * @returns {number} 1-based token index
 */
export function tokenIndex(alphabet, symbol) {
  const idx = alphabet.indexOf(symbol);
  if (idx < 0) throw new Error(`Unknown symbol "${symbol}". Valid symbols: ${alphabet.slice(1).join(', ')}`);
  return idx;
}

/**
 * Convert a string/array of symbols to Uint32Array of token indices.
 * @param {string|string[]} seq - Sequence of symbols
 * @param {string[]} alphabet - Token alphabet (index 0 = null)
 * @returns {Uint32Array} 1-based token indices
 */
export function tokenize(seq, alphabet) {
  const symbols = typeof seq === 'string' ? seq.split('') : seq;
  const result = new Uint32Array(symbols.length);
  for (let i = 0; i < symbols.length; i++) {
    result[i] = tokenIndex(alphabet, symbols[i]);
  }
  return result;
}

/**
 * Extract the silent transition submatrix logTrans[0][0][src][dst].
 * @param {Float64Array} logTrans
 * @param {number} nOut
 * @param {number} S
 * @returns {Float64Array} S*S matrix
 */
export function extractSilent(logTrans, nOut, S) {
  const silent = new Float64Array(S * S);
  const base = 0; // inTok=0, outTok=0
  for (let i = 0; i < S * S; i++) {
    silent[i] = logTrans[base + i];
  }
  return silent;
}

File: rust/fused-plan7/src/machine_prep.rs
//! Machine preparation: JSON transducer + params -> SPARSE log-transition lists.
//!
//! Faithful port of `js/webgpu/internal/machine-prep.mjs`, EXCEPT the transition
//! tensor is stored sparsely instead of as a dense `[n_in * n_out * S * S]`
//! array. The JS reference iterates the dense `S*S` block for every emission;
//! for a real transducer (e.g. `prot2dna` with ~132 states) that is
//! `O(L * K * S^2)`. We instead store, per `(in_tok, out_tok)`, the list of
//! `(src, dst, log_weight)` actually present, giving `O(L * K * nnz)`.
//!
//! Skipping `NEG_INF` (absent) entries is *exactly* behaviour-preserving: the
//! semiring reduce (logsumexp / max) treats `NEG_INF` as its identity, so a
//! sparse fold over present entries equals the dense reduce over all entries.
//!
//! Token convention (identical to JS): token 0 is the null (epsilon/gap) token;
//! real tokens are 1-based indices into the sorted alphabet.

use serde_json::Value;
use std::collections::BTreeMap;
use std::collections::HashMap;

use crate::logmath::{logaddexp, NEG_INF};

/// One sparse transition: `value[dst] = plus(value[dst], v[src] + log_weight)`.
#[derive(Clone, Copy, Debug)]
pub struct SparseEdge {
    pub src: u32,
    pub dst: u32,
    pub log_weight: f64,
}

/// A prepared machine with sparse log-transition lists.
///
/// Mirrors `PreparedMachine` but stores transitions sparsely. The dense layout
/// it replaces was `logTrans[in*nOut*S*S + out*S*S + src*S + dst]`.
#[derive(Clone, Debug)]
pub struct PreparedMachine {
    pub n_states: usize,
    pub n_input_tokens: usize,
    pub n_output_tokens: usize,
    pub input_alphabet: Vec<String>,
    pub output_alphabet: Vec<String>,
    /// Sparse edges keyed by `(in_tok, out_tok)`. Within each list, edges are
    /// sorted by `dst` ascending then `src` ascending. The DP consumes them
    /// grouped by `dst`: for each `dst` it folds `v[src] + log_weight` over the
    /// contiguous run of edges with that `dst`, in `src`-ascending order. That is
    /// exactly the per-column reduce the dense JS scan performs
    /// (`reduce_src(v[src] + trans[src*S+dst])`), so the logsumexp accumulation is
    /// bit-identical. (Absent `src` entries are NEG_INF = reduce identity, so
    /// skipping them changes nothing.)
    pub edges: HashMap<(usize, usize), Vec<SparseEdge>>,
    /// The silent block `(in=0, out=0)` as a dense `S*S` matrix in row-major
    /// `[src*S + dst]` order (NEG_INF where absent). The JS `td_silent` is used
    /// densely (fixed-point iteration touches every entry), so we keep it dense
    /// too for an exact match.
    pub silent_dense: Vec<f64>,
}

impl PreparedMachine {
    /// Borrow the sparse edge list for `(in_tok, out_tok)`, or an empty slice.
    #[inline]
    pub fn edges_for(&self, in_tok: usize, out_tok: usize) -> &[SparseEdge] {
        match self.edges.get(&(in_tok, out_tok)) {
            Some(v) => v.as_slice(),
            None => &[],
        }
    }
}

/// Evaluate a weight expression to f64 given params + defs.
///
/// Faithful port of `evaluateWeight` (which mirrors `weight.py:evaluate`).
/// Supports number / bool / string (param or def lookup) / `{*,+,-,/,pow,log,exp,not}`.
pub fn evaluate_weight(
    w: &Value,
    params: &HashMap<String, f64>,
    defs: &HashMap<String, Value>,
) -> Result<f64, String> {
    match w {
        Value::Null => Ok(0.0),
        Value::Bool(b) => Ok(if *b { 1.0 } else { 0.0 }),
        Value::Number(n) => n
            .as_f64()
            .ok_or_else(|| "weight: non-finite number".to_string()),
        Value::String(s) => {
            if let Some(v) = params.get(s) {
                Ok(*v)
            } else if let Some(d) = defs.get(s) {
                evaluate_weight(d, params, defs)
            } else {
                Err(format!(
                    "Unknown parameter \"{}\". Pass it in the params map.",
                    s
                ))
            }
        }
        Value::Object(map) => {
            // Binary operators take a 2-element array.
            let bin = |key: &str| -> Option<(&Value, &Value)> {
                map.get(key).and_then(|v| v.as_array()).and_then(|a| {
                    if a.len() == 2 {
                        Some((&a[0], &a[1]))
                    } else {
                        None
                    }
                })
            };
            if let Some((a, b)) = bin("*") {
                return Ok(evaluate_weight(a, params, defs)? * evaluate_weight(b, params, defs)?);
            }
            if let Some((a, b)) = bin("+") {
                return Ok(evaluate_weight(a, params, defs)? + evaluate_weight(b, params, defs)?);
            }
            if let Some((a, b)) = bin("-") {
                return Ok(evaluate_weight(a, params, defs)? - evaluate_weight(b, params, defs)?);
            }
            if let Some((a, b)) = bin("/") {
                return Ok(evaluate_weight(a, params, defs)? / evaluate_weight(b, params, defs)?);
            }
            if let Some((base, exp)) = bin("pow") {
                return Ok(evaluate_weight(base, params, defs)?
                    .powf(evaluate_weight(exp, params, defs)?));
            }
            if let Some(v) = map.get("log") {
                return Ok(evaluate_weight(v, params, defs)?.ln());
            }
            if let Some(v) = map.get("exp") {
                return Ok(evaluate_weight(v, params, defs)?.exp());
            }
            if let Some(v) = map.get("not") {
                return Ok(1.0 - evaluate_weight(v, params, defs)?);
            }
            Err(format!(
                "Unsupported weight operator \"{}\". Supported: *, +, -, /, pow, log, exp, not",
                map.keys().cloned().collect::<Vec<_>>().join(", ")
            ))
        }
        _ => Err("Unsupported weight expression type".to_string()),
    }
}

/// Canonicalize a state id/`to` reference to a lookup key, matching the JS
/// `Array.isArray(name) ? JSON.stringify(name) : name`.
fn ref_key(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        Value::Array(_) | Value::Object(_) => serde_json::to_string(v).unwrap_or_default(),
        Value::Number(n) => n.to_string(),
        _ => serde_json::to_string(v).unwrap_or_default(),
    }
}

/// Build a sorted token alphabet (index 0 = null/epsilon), mirroring
/// `buildAlphabet`: collect every non-empty `direction` token over all
/// transitions, sort lexicographically, prepend "".
fn build_alphabet(states: &[Value], direction: &str) -> Vec<String> {
    let mut set = std::collections::BTreeSet::new();
    for st in states {
        if let Some(trans) = st.get("trans").and_then(|t| t.as_array()) {
            for t in trans {
                if let Some(tok) = t.get(direction).and_then(|x| x.as_str()) {
                    if !tok.is_empty() {
                        set.insert(tok.to_string());
                    }
                }
            }
        }
    }
    let mut out = vec![String::new()];
    out.extend(set);
    out
}

/// Prepare a machine JSON + params into sparse log-transition lists.
///
/// Faithful port of `prepareMachine` but sparse. Duplicate transitions for the
/// same `(in, out, src, dst)` are combined with `logaddexp`, exactly as the JS.
pub fn prepare_machine(
    machine_json: &Value,
    params: &HashMap<String, f64>,
) -> Result<PreparedMachine, String> {
    let states = machine_json
        .get("state")
        .and_then(|s| s.as_array())
        .ok_or_else(|| "machine JSON has no \"state\" array".to_string())?;

    let defs: HashMap<String, Value> = match machine_json.get("defs").and_then(|d| d.as_object()) {
        Some(obj) => obj.iter().map(|(k, v)| (k.clone(), v.clone())).collect(),
        None => HashMap::new(),
    };

    let s = states.len();
    let input_alphabet = build_alphabet(states, "in");
    let output_alphabet = build_alphabet(states, "out");
    let n_in = input_alphabet.len();
    let n_out = output_alphabet.len();

    // token -> index maps
    let in_tok_idx: HashMap<&str, usize> = input_alphabet
        .iter()
        .enumerate()
        .map(|(i, t)| (t.as_str(), i))
        .collect();
    let out_tok_idx: HashMap<&str, usize> = output_alphabet
        .iter()
        .enumerate()
        .map(|(i, t)| (t.as_str(), i))
        .collect();

    // Resolve state name references to indices (id can be string/array/number;
    // numeric index always maps to itself).
    let mut name_to_idx: HashMap<String, usize> = HashMap::new();
    for (i, st) in states.iter().enumerate() {
        if let Some(id) = st.get("id") {
            name_to_idx.insert(ref_key(id), i);
        }
        name_to_idx.insert(i.to_string(), i);
    }

    let resolve_dest = |dest: &Value| -> Result<usize, String> {
        if let Some(n) = dest.as_u64() {
            return Ok(n as usize);
        }
        let key = ref_key(dest);
        name_to_idx
            .get(&key)
            .copied()
            .ok_or_else(|| format!("Transition references unknown state \"{}\".", key))
    };

    // Accumulate into a (in,out,src,dst) -> log_weight map so duplicate
    // transitions are logaddexp-combined (matching the JS dense fill).
    // BTreeMap keeps deterministic iteration; we re-sort per (in,out) list below.
    let mut acc: BTreeMap<(usize, usize, u32, u32), f64> = BTreeMap::new();

    for (src, st) in states.iter().enumerate() {
        if let Some(trans) = st.get("trans").and_then(|t| t.as_array()) {
            for t in trans {
                let to = t
                    .get("to")
                    .ok_or_else(|| "transition missing \"to\"".to_string())?;
                let dst = resolve_dest(to)?;
                let in_idx = match t.get("in").and_then(|x| x.as_str()) {
                    Some(tok) if !tok.is_empty() => *in_tok_idx
                        .get(tok)
                        .ok_or_else(|| format!("unknown input token \"{}\"", tok))?,
                    _ => 0,
                };
                let out_idx = match t.get("out").and_then(|x| x.as_str()) {
                    Some(tok) if !tok.is_empty() => *out_tok_idx
                        .get(tok)
                        .ok_or_else(|| format!("unknown output token \"{}\"", tok))?,
                    _ => 0,
                };
                let weight_expr = t.get("weight").cloned().unwrap_or(Value::from(1));
                let log_weight = evaluate_weight(&weight_expr, params, &defs)?.ln();
                let key = (in_idx, out_idx, src as u32, dst as u32);
                acc.entry(key)
                    .and_modify(|w| *w = logaddexp(*w, log_weight))
                    .or_insert(log_weight);
            }
        }
    }

    // Materialize sparse edge lists per (in,out), then sort each list by
    // (dst, src) so the DP can fold per-dst columns in src-ascending order
    // (bit-identical to the dense per-column reduce; see the `edges` doc).
    let mut edges: HashMap<(usize, usize), Vec<SparseEdge>> = HashMap::new();
    let mut silent_dense = vec![NEG_INF; s * s];
    for ((in_idx, out_idx, src, dst), lw) in acc.into_iter() {
        edges
            .entry((in_idx, out_idx))
            .or_default()
            .push(SparseEdge {
                src,
                dst,
                log_weight: lw,
            });
        if in_idx == 0 && out_idx == 0 {
            silent_dense[src as usize * s + dst as usize] = lw;
        }
    }
    for list in edges.values_mut() {
        list.sort_by(|a, b| a.dst.cmp(&b.dst).then(a.src.cmp(&b.src)));
    }

    Ok(PreparedMachine {
        n_states: s,
        n_input_tokens: n_in,
        n_output_tokens: n_out,
        input_alphabet,
        output_alphabet,
        edges,
        silent_dense,
    })
}

/// 1-based token index for `symbol` in `alphabet` (index 0 = null).
/// Mirrors `tokenIndex`.
pub fn token_index(alphabet: &[String], symbol: &str) -> Result<u32, String> {
    alphabet
        .iter()
        .position(|a| a == symbol)
        .map(|p| p as u32)
        .ok_or_else(|| {
            format!(
                "Unknown symbol \"{}\". Valid symbols: {}",
                symbol,
                alphabet[1..].join(", ")
            )
        })
}

/// Convert a string of single-character symbols to 1-based token indices.
/// Mirrors `tokenize` for the string-input case.
pub fn tokenize(seq: &str, alphabet: &[String]) -> Result<Vec<u32>, String> {
    seq.chars()
        .map(|c| token_index(alphabet, &c.to_string()))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::logmath::Semiring;
    use serde_json::json;

    #[test]
    fn evaluate_weight_operators() {
        let p: HashMap<String, f64> = [("x".to_string(), 2.0), ("y".to_string(), 3.0)]
            .into_iter()
            .collect();
        let d: HashMap<String, Value> = HashMap::new();
        assert_eq!(evaluate_weight(&json!({"*": ["x", "y"]}), &p, &d).unwrap(), 6.0);
        assert_eq!(evaluate_weight(&json!({"+": ["x", "y"]}), &p, &d).unwrap(), 5.0);
        assert_eq!(evaluate_weight(&json!({"-": ["y", "x"]}), &p, &d).unwrap(), 1.0);
        assert_eq!(evaluate_weight(&json!({"/": ["y", "x"]}), &p, &d).unwrap(), 1.5);
        assert_eq!(evaluate_weight(&json!({"pow": ["x", "y"]}), &p, &d).unwrap(), 8.0);
        assert_eq!(evaluate_weight(&json!({"not": 0.25}), &p, &d).unwrap(), 0.75);
        assert!((evaluate_weight(&json!({"exp": {"log": "x"}}), &p, &d).unwrap() - 2.0).abs() < 1e-12);
        assert_eq!(evaluate_weight(&json!(1), &p, &d).unwrap(), 1.0);
        assert_eq!(evaluate_weight(&Value::Null, &p, &d).unwrap(), 0.0);
        assert!(evaluate_weight(&json!("missing"), &p, &d).is_err());
    }

    #[test]
    fn defs_lookup_resolves() {
        let machine = json!({
            "defs": { "half": 0.5 },
            "state": [{ "id": "S", "trans": [{ "to": "S", "in": "a", "out": "a", "weight": "half" }] }]
        });
        let prep = prepare_machine(&machine, &HashMap::new()).unwrap();
        // log(0.5) on the single (in=a,out=a) edge.
        let e = prep.edges_for(1, 1);
        assert_eq!(e.len(), 1);
        assert!((e[0].log_weight - 0.5_f64.ln()).abs() < 1e-12);
    }

    /// The core correctness claim of the sparse representation: folding the
    /// sparse per-`dst` edge lists equals reducing the equivalent DENSE `S*S`
    /// matrix (NEG_INF where absent). Verified for both semirings on a small
    /// multi-state transducer with a random-ish source vector.
    #[test]
    fn sparse_fold_equals_dense_reduce() {
        // 3-state machine, a single (in=a,out=b) block plus some silent edges.
        let machine = json!({
            "state": [
                { "id": 0, "trans": [
                    { "to": 1, "in": "a", "out": "b", "weight": 0.5 },
                    { "to": 2, "in": "a", "out": "b", "weight": 0.25 },
                    { "to": 0 } // silent self-loop weight 1
                ]},
                { "id": 1, "trans": [
                    { "to": 2, "in": "a", "out": "b", "weight": 0.1 },
                    { "to": 1, "in": "a", "out": "b", "weight": 0.7 }
                ]},
                { "id": 2, "trans": [
                    { "to": 0, "in": "a", "out": "b", "weight": 0.9 }
                ]}
            ]
        });
        let prep = prepare_machine(&machine, &HashMap::new()).unwrap();
        let s = prep.n_states;
        assert_eq!(s, 3);
        // tokens: in alphabet ["","a"], out ["","b"] -> a=1, b=1.
        let (in_tok, out_tok) = (1usize, 1usize);

        // Build the dense block for (in_tok,out_tok).
        let mut dense = vec![NEG_INF; s * s];
        for e in prep.edges_for(in_tok, out_tok) {
            dense[e.src as usize * s + e.dst as usize] = e.log_weight;
        }

        let v = [-0.3_f64, -1.2, -2.5];
        for sem in [Semiring::LogSumExp, Semiring::MaxPlus] {
            // Dense reduce per column (the JS tdMatvec).
            let dense_out: Vec<f64> = (0..s)
                .map(|dst| sem.reduce((0..s).map(|src| v[src] + dense[src * s + dst])))
                .collect();
            // Sparse fold per dst-group (what td_emit/td_delete do).
            let edges = prep.edges_for(in_tok, out_tok);
            let mut sparse_out = vec![NEG_INF; s];
            let mut i = 0;
            while i < edges.len() {
                let dst = edges[i].dst as usize;
                let mut j = i;
                while j < edges.len() && edges[j].dst as usize == dst {
                    j += 1;
                }
                sparse_out[dst] =
                    sem.reduce(edges[i..j].iter().map(|e| v[e.src as usize] + e.log_weight));
                i = j;
            }
            for st in 0..s {
                let (a, b) = (dense_out[st], sparse_out[st]);
                assert!(
                    (a == NEG_INF && b == NEG_INF) || (a - b).abs() < 1e-15,
                    "{:?} col {}: dense {} vs sparse {}",
                    sem,
                    st,
                    a,
                    b
                );
            }
        }
    }

    #[test]
    fn duplicate_transitions_logaddexp_combine() {
        // Two identical (in=a,out=b,src=0,dst=0) edges should combine via logaddexp.
        let machine = json!({
            "state": [{ "id": 0, "trans": [
                { "to": 0, "in": "a", "out": "b", "weight": 0.3 },
                { "to": 0, "in": "a", "out": "b", "weight": 0.4 }
            ]}]
        });
        let prep = prepare_machine(&machine, &HashMap::new()).unwrap();
        let e = prep.edges_for(1, 1);
        assert_eq!(e.len(), 1, "duplicate edges merge into one");
        let expected = crate::logmath::logaddexp(0.3_f64.ln(), 0.4_f64.ln());
        assert!((e[0].log_weight - expected).abs() < 1e-15);
    }
}
